File: src/limits.cpp
// This file is part of the NewtonFractal project.
// Copyright (C) 2019 Christian Bauer and Timon Foehl
// License: GNU General Public License version 3 or later,
// see the file LICENSE in the main directory.

#include "limits.h"
#include "defaults.h"

Limits::Limits(bool original) :
	left_(-1.0),
	right_(1.0),
	top_(1.0),
	bottom_(-1.0),
	original_(nullptr)
{
	// Create original
	if (!original) {
		original_ = new Limits(true);
	}
}

Limits::~Limits()
{
	// Delete original
	if (original_ != nullptr) {
		delete original_;
	}
}

Limits &Limits::operator=(const Limits &other)
{
	// Copy limits
	set(other.left(), other.right(), other.top(), other.bottom());

	// Deep copy original limits
	if (original_ != nullptr) {
		setOriginal(
			other.original()->left(),
			other.original()->right(),
			other.original()->top(),
			other.original()->bottom());
	}
	return *this;
}

bool Limits::operator==(const Limits &other) const
{
	// Check if limits are the same
	return (
		left_ == other.left() &&
		right_ == other.right() &&
		top_ == other.top() &&
		bottom_ == other.bottom()
	);
}

bool Limits::operator!=(const Limits &other) const
{
	// Check if limits are not the same
	return (
		left_ != other.left() ||
		right_ != other.right() ||
		top_ != other.top() ||
		bottom_ != other.bottom()
	);
}

void Limits::move(QPoint distance, const QSize &ref)
{
	// Move limits by distance
	double dx = distance.x() * (right_ - left_) / (ref.width() - 1);
	double dy = distance.y() * (bottom_ - top_) / (ref.height() - 1);
	left_ += dx;
	right_ += dx;
	top_ += dy;
	bottom_ += dy;

	// Move original limits
	if (original_ != nullptr) {
		original_->move(distance, ref);
	}
}

void Limits::zoom(bool in, double xw, double yw)
{
	// Zoom limits in / out
	double zoom = in ? -nf::ZMF : nf::ZMF;
	double wZoom = (right_ - left_) * zoom;
	double hZoom = (top_ - bottom_) * zoom;
	left_ -= xw * wZoom;
	right_ += (1.0 - xw) * wZoom;
	top_ += yw * hZoom;
	bottom_ -= (1.0 - yw) * hZoom;
}

void Limits::reset(QSize size)
{
	// Reset limits to match size
	left_ = -nf::DSF * size.width();
	right_ = nf::DSF * size.width();
	top_ = nf::DSF * size.height();
	bottom_ = -nf::DSF * size.height();

	// Reset original limits
	if (original_ != nullptr) {
		original_->reset(size);
	}
}

void Limits::resize(QSize delta)
{
	// Resize limits
	right_ += nf::DSF * delta.width();
	left_ -= nf::DSF * delta.width();
	top_ += nf::DSF * delta.height();
	bottom_ -= nf::DSF * delta.height();

	// Resize original limits
	if (original_ != nullptr) {
		original_->resize(delta);
	}
}

void Limits::set(double left, double right, double top, double bottom)
{
	// Set limits
	left_ = left;
	right_ = right;
	top_ = top;
	bottom_ = bottom;
}

void Limits::setOriginal(double left, double right, double top, double bottom)
{
	// Set original limits
	original_->set(left, right, top, bottom);
}

double Limits::width() const
{
	// Return width
	return right_ - left_;
}

double Limits::height() const
{
	// Return height
	return top_ - bottom_;
}

double Limits::left() const
{
	// Return left limit
	return left_;
}

double Limits::right() const
{
	// Return right limit
	return right_;
}

double Limits::top() const
{
	// Return top limit
	return top_;
}

double Limits::bottom() const
{
	// Return bottom limit
	return bottom_;
}

QVector4D Limits::vec4() const
{
	// Return limits as vec4
	return QVector4D(top_, right_, bottom_, left_);
}

double Limits::zoomFactor() const
{
	// Return zoomFactor
	return original_->width() / width();
}

void Limits::setZoomFactor(double zoomFactor)
{
	// Set zoomFactor
	double w2 = 0.5 * original_->width() / zoomFactor;
	double h2 = 0.5 * original_->height() / zoomFactor;
	double xMid = 0.5 * right_ + 0.5 * left_;
	double yMid = 0.5 * top_ + 0.5 * bottom_;
	right_ = xMid + w2;
	left_ = xMid - w2;
	top_ = yMid + h2;
	bottom_ = yMid - h2;
}

const Limits *Limits::original() const
{
	// Return pointer to original limits
	return original_;
}

File: docs/limits.js
// This file is part of the NewtonFractal project.
// Copyright (C) 2019 Christian Bauer and Timon Foehl
// License: GNU General Public License version 3 or later,
// see the file LICENSE in the main directory.

class Limits {
	constructor(original = false) {
		this.top = 1.0;
		this.right = 1.0;
		this.bottom = -1.0;
		this.left = -1.0;
		if (!original) {
			this.original = new Limits(true);
		} else this.original = null;
	}

	width() { 
		return this.right - this.left; 
	}

	height() { 
		return this.top - this.bottom; 
	}

	vec4() {
		return [this.top, this.right, this.bottom, this.left];
	}

	zoomFactor() {
		return this.original.width() / this.width();
	}

	setZoomFactor(zf) {
		let w2 = 0.5 * this.original.width() / zf;
		let h2 = 0.5 * this.original.height() / zf;
		let xMid = 0.5 * this.right + 0.5 * this.left;
		let yMid = 0.5 * this.top + 0.5 * this.bottom;
		this.top = yMid + h2;
		this.right = xMid + w2;
		this.bottom = yMid - h2;
		this.left = xMid - w2;
	}
	
	zoom(zoomin, xw, yw) {
		let zoom = zoomin ? -nf.ZMF : nf.ZMF;
		let wZoom = this.width() * zoom;
		let hZoom = this.height() * zoom;
		this.top += yw * hZoom;
		this.right += (1.0 - xw) * wZoom;
		this.bottom -= (1.0 - yw) * hZoom;
		this.left -= xw * wZoom;
	}

	move(x, y) {
		let dx = x * this.width() / (canvas.width - 1);
		let dy = y * -this.height() / (canvas.height - 1);
		this.top += dy;
		this.right += dx;
		this.bottom += dy;
		this.left += dx;
		if (this.original != null) {
			this.original.move(x, y);
		}
	}

	resize(x, y) {
		this.top += nf.DSF * y;
		this.right += nf.DSF * x;
		this.bottom -= nf.DSF * y;
		this.left -= nf.DSF * x;
		if (this.original != null) {
			this.original.resize(x, y);
		}
	}

	reset() {
		this.top = nf.DSF * canvas.height;
		this.right = nf.DSF * canvas.width;
		this.bottom = -nf.DSF * canvas.height;
		this.left = -nf.DSF * canvas.width;
		if (this.original != null) {
			this.original.reset();
		}
	}
}